[mono] merge wasm-threading-eventpipe into main

lambdageek opened the pull request:
Merge initial work on multi-threaded WebAssembly.

The normal wasm build is single-threaded.  There should be no functional changes to its behavior.

To enable a multi-threaded build pass `/p:WasmEnableThreads=true`.  See `src/mono/wasm/threads.md` for details.

The big changes are:
1. The normal ref assemblies related to threading retain `[UnsupportedOSPlatform("browser")]` attributes for various threading-related functions
2. In System.Private.CoreLib, the `[UnsupportedOSPlatform]` attributes are removed, and functions that used to always throw PNSE nwo do a runtime check using `System.Threading.Thread.IsThreadStartSupported` to check if threading is enabled in the current build. 
3. A new nuget `Microsoft.NET.WebAssembly.Threading` is created.  It contains experimental ref assemblies without the `[UnsupportedOSPlatform]` attributes.  The intention is that code opting into experimenting with multithreading will include this nuget by setting some property that will be used by the wasm MSBuild SDK to pick a multi-threaded runtime build and configure things appropriately.  (The SDK updates don't exist yet).
4. In the multi-threaded runtime we don't use Emscripten's "main thread" option (ie: the browser thread is the main one); we also continue to run certain runtime-internal jobs (finalizers, GC pump) on the main thread even in the multi-threaded runtime.

Remaining work is tracked in the related issue https://github.com/dotnet/runtime/issues/68162

radekdoulik committed:
Initial changes for emscripten 2.0.34

radekdoulik committed:
Use emcc-link.rsp in build targets

radekdoulik committed:
Use updated docker images

Commit:
Fix compiler warnings

Co-authored-by: Zoltan Varga <vargaz@gmail.com>

radekdoulik committed:
Put `--profiling-funcs` to `_EmccLinkFlags`

radekdoulik committed:
Fix build

    src/mono/mono/mini/mini-runtime.c:3407:25: error: ‘invoke’ undeclared (first use in this function); did you mean ‘revoke’?
       3407 |                         invoke = mono_marshal_get_runtime_invoke_dynamic ();

radekdoulik committed:
Add shell to the environment

Environment setting https://github.com/emscripten-core/emscripten/blob/2.0.34/src/settings.js#L616-L641

From emscripten 2.0.25 release notes

    - Support for the 'shell' environment is now disabled by default.  Running under
      `d8`, `js`, or `jsc` is not something that most emscripten users ever want to
      do, so including the support code is, more often than not, unnecessary.  Users
      who want shell support can enable it by including 'shell' in `-s ENVIRONMENT`
      (#14535).

Example of the the size increase for bench sample:

    -a---          12/10/2021  3:35 PM         382113 dotnet.js
    -a---          12/13/2021 10:37 AM         383589 dotnet.js

radekdoulik committed:
Merge remote-tracking branch 'origin/main' into pr-emscripten-34-bump

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-emscripten-34-bump

radekdoulik committed:
Add emcc-link.rsp to PlatformManifestFileEntry

radekdoulik committed:
Feedback

https://github.com/emscripten-core/emscripten/blob/2fda25eea756c78c8cb024aa5b6c2b188bf7990f/src/settings.js#L1173-L1176

-s EXPORT_ES6 is link option

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-emscripten-34-bump

radekdoulik committed:
Bump emscripten version

radekdoulik committed:
Bump llvm package version and use its libclang

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-emscripten-34-bump

radekdoulik committed:
Use newer docker images with emscripten 3.1.1

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-emscripten-34-bump

radekdoulik committed:
Remove unused variable

Commit:
Merge remote-tracking branch 'upstream/main' into wasm-threaded-build

Commit:
Add runtime support for threads in the wasm build

To enable, pass `/p:WasmEnableThreads` when building the runtime

./build.sh -Subset mono+libs -os Browser -arch wasm /p:WasmEnableThreads=true

kg committed:
Prevent runtime from starting twice when loaded in a web worker

kg committed:
Automatically populate the emscripten mainScriptUrlOrBlob property so that worker initialization can find dotnet.js

kg committed:
Add compatibility shim so that emscripten's generated worker.js can properly get a Module instance, since we broke the API

kg committed:
Checkpoint

kg committed:
Add comments and fix a typo

kg committed:
Introduce 'MonoObjectRef' ts type. Migrate mono_string_intern to not have a retval

kg committed:
Checkpoint (strings are broken for some reason)

kg committed:
Fix string interning

kg committed:
Checkpoint

kg committed:
Migrate ObjectToString and GetDateValue

kg committed:
Checkpoint gc safe/unsafe region work

kg committed:
Checkpoint

kg committed:
More ref conversion

kg committed:
Checkpoint (broken?)

kg committed:
Fix missing method

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint (broken)

kg committed:
Checkpoint (broken)

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint (broken)

kg committed:
Fix incorrect signatures

kg committed:
Checkpoint

kg committed:
Fix lint

kg committed:
Add new 'R' signature char for 'ref object'

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Remove AddEventListener and RemoveEventListener
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
eslint fixes

kg committed:
Checkpoint

kg committed:
Update call_method signature to avoid passing raw object pointers

kg committed:
Ref-ify one websocket API and fix types and incorrect rooting of two others

kg committed:
Checkpoint

kg committed:
Deprecation metadata

kg committed:
More ref

kg committed:
More ref

kg committed:
Remove outdated comments

kg committed:
Convert typed_array_new to ref

kg committed:
Add volatile modifiers, satisfy eslint

kg committed:
Update src/mono/wasm/runtime/corebindings.c

Co-authored-by: Aleksey Kliger (λgeek) <akliger@gmail.com>

Commit:
Bring back threadpool and add some tracing in diagnostics

Commit:
Missing conflict

Commit:
Merge remote-tracking branch 'upstream/main' into wasm-eventpipe

Commit:
Fix build, set coop gc, and always copy dotnet.worker.js when it's around for apps

Commit:
Disable sample profiler, add some functions that were missing from katelyn's PR.

Commit:
Add safepoint around ep_rt_wait_event_set.  Tweak sample to explicitly exit wasm in order to properly flush event to the nettrace file (w/ rundown events).

lambdageek committed:
[gc] Start the GC Finalizer thread on threaded WASM

lambdageek committed:
[mono] add GC Unsafe in mono_assembly_load; remove in EventPipe

Remove GC Unsafe hack in ep_rt_wait_event_set

lambdageek committed:
Add mono_threads_wasm_async_run_in_main_thread; do background work on main

Don't start a finalizer thread

Queue all background work to run on the main thread

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-threading-eventpipe

lambdageek committed:
post-merge cleanup: delete duplicated definitions

lambdageek committed:
[sample] Env vars should be stringy

lambdageek committed:
updated dotnet.d.ts

lambdageek committed:
Add a System.Threading.Thread.WebAssembly.Threading ref assembly

lambdageek committed:
Update the browser sample to use System.Threading.Thread.WebAssembly.Threading

lambdageek committed:
[mono] Fix non-threaded wasm build

lambdageek committed:
what is oob-all for? don't add thread wasm ref assembly there

Does System.Threading.Thread.WebAssembly.Threading need to be in any of the
.proj files in src/libraries?

lambdageek committed:
Rationalize System.Threading.Thread build

In CoreLib, never add the [UnsupportedOSPlatform("browser")] attribute. Rely on
runtime checks (`ThrowIfNoThreadStart()`).
In System.Threading.Thread ref assembly, always add the unsupported platform
attribute.  Add mismatches to ApiCompat baseline.
In System.Threading.Thread.WebAssembly.Threading don't add the attributes, and
also set DisablePackageBaselineValidation to prevent Restore from looking for a
System.Threading.Thread.WebAssembly.Threading nuget (fixes in-tree
ProjectReferences for testing)

lambdageek committed:
only turn on analyzers for the browser sample

not all wasm samples

lambdageek committed:
Make a single Microsoft.NET.WebAssembly.Threading nupkg

that holds all the special ref assemblies

lambdageek committed:
works: sample has ProjectReference to the Microsoft.NET.WebAssembly.Threading.proj

lambdageek committed:
Add System.Threading.ThreadPool.WebAssembly.Threading ref assembly

lambdageek committed:
ThreadPool: throw PNSE if no thread start

lambdageek committed:
Update wasm threads.md

lambdageek committed:
apicompat: correct warnings for System.Threading.ThreadPool

lambdageek committed:
Add dotnet.worker.js to the runtime pack; update PlatformManifestFileEntry

lambdageek committed:
[wasm] startup: detect Blazor dotnet.[version].[hash].js location

Blazor injects a `<link rel="modulepreload" />` element into the header when it
boots; detect it and extract the URL of the script.

This is needed by Emscripten's dotnet.worker.js to run WorkerGlobalScope.importScripts

lambdageek committed:
one more fix to Microsoft.NET.WebAssembly.Threading

Seems to be necessary in order for the resulting .nupkg not to reference
non-existent nugets for the ProjectReferences

lambdageek committed:
rename sample to browser-mt-eventpipe

lambdageek committed:
bring back unmodified browser sample

The multithreading sample is browser-mt-eventpipe

lambdageek committed:
update browser-mt-eventpipe sample to use ref assembly refs

Referencing the rollup Microsoft.NET.WebAssembly.Threading.proj doesn't
work (it ends up bundling the ref assemblies into the publish dir and breaking
the app)

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-threading-eventpipe

lambdageek committed:
Use correct ifdef in AppContext.AnyOS.cs

lambdageek committed:
[build] support WasmEnableThreads and WasmEnablePerfTracing

These toplevel options either turn on multi-threading in general, or turn on
multithreading only for eventpipe internals.

For libraries these define `FeatureWasmThreads` and `FeatureWasmPerfTracing`
properties and the `FEATURE_WASM_THREADS` and `FEATURE_WASM_PERFTRACING`
compiler constants.

For the native code, they control `DISABLE_THREADS` and
`DISABLE_WASM_USER_THREADS` cmake and preprocessor settings.

lambdageek committed:
Only add the portable threadpool on wasm if threading is enabled

lambdageek committed:
rename browser-mt-eventpipe csproj and main assembly

Give it a unique name distinct from Wasm.Browser.CJS.Sample.csproj

lambdageek committed:
fix /p:WasmEnableThreads=false build after merge

lambdageek committed:
now fix the WasmEnableThreads=true build

lambdageek committed:
we need two ThreadPoolBoundHandle implementation stubs

one for PortableThreadPool when threads are enabled, one for single-threaded wasm

lambdageek committed:
Add a System.Diagnostics.Tracing ref assembly gated by FeatureWasmPerfTracing

lambdageek committed:
[eventpipe] use the correct cmake option name

see src/mono/mono.proj

lambdageek committed:
revert debug printf and commented out code

lambdageek committed:
turn off additional logging

lambdageek committed:
hack: set FeatureWasmPerfTracing in the browser-mt-eventpipe sample

It would be better to drive this (the inclusion of the tracing runtime
component) from a user-visible flag. not the FeatureWasmPerfTracing
implementation detail

lambdageek commented:
/azp run runtime-wasm

vargaz commented on src/mono/mono.proj:
```diff
@@ -321,17 +325,25 @@
     <ItemGroup Condition="'$(TargetsBrowser)' == 'true'">
-      <_MonoCMakeArgs Include="-DENABLE_MINIMAL=jit,sgen_major_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,sgen_debug_helpers,sgen_binary_protocol,logging,interpreter,threads,qcalls$(_MonoMinimal)"/>
+      <_MonoCMakeArgs Include="-DENABLE_MINIMAL=jit,sgen_major_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,sgen_debug_helpers,sgen_binary_protocol,logging,interpreter,qcalls$(_MonoMinimal)"/>
       <_MonoCMakeArgs Include="-DENABLE_INTERP_LIB=1"/>
```
Note that this will lead to some size increase since enable-minimal=threads linked out mutexes, tls, etc.

lambdageek replied:
It's not adding it unconditionally. It's added to `_MonoMinimal` for single-threaded builds.

vargaz commented on src/mono/mono/utils/memfuncs.c:
```diff
@@ -53,7 +53,7 @@
 #define align_up(ptr) ((void*) ((_toi(ptr) + ptr_mask) & ~ptr_mask))
-#if SIZEOF_VOID_P == 4
+#if SIZEOF_VOID_P == 4 || SIZEOF_VOID_P == 7
 #define bytes_to_words(n)	((size_t)(n) >> 2)
```
Unrelated ?

lambdageek replied:
@steveisok Do you remember what this is for?

steveisok replied:
I forgot to take it out.  Early on I was messing with the build and sometimes `SIZEOF_VOID_P` would be 7 for unknown reasons. Super weird, so we should look into how it can happen.

vargaz commented on src/mono/mono/utils/mono-threads-wasm.c:
```diff
@@ -110,6 +115,7 @@ mono_native_thread_id_equals (MonoNativeThreadId id1, MonoNativeThreadId id2)
 }
 
+static int pthread_warning_count = 0;
 
```
Debug code?

lambdageek committed:
remove unused variable

lambdageek commented:
/azp run runtime-wasm

kg commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+      <_FileVersionMin>$(FileVersion.Split('.')[1])</_FileVersionMin>
+      <_FileVersionBld>$(FileVersion.Split('.')[2])</_FileVersionBld>
+      <_FileVersionRev>$(FileVersion.Split('.')[3])</_FileVersionRev>
+      <FileVersion>$(_FileVersionMaj).$([MSBuild]::Add($(_FileVersionMin), 100)).$(_FileVersionBld).$(_FileVersionRev)</FileVersion>
```
I assume you copied and pasted this from somewhere else but I hate seeing it without a comment that explains why we even do it

kg commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -2326,7 +2326,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Threading\ThreadPool.Portable.cs" Condition="'$(FeatureCoreCLR)' != 'true'" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Threading\ThreadPool.Portable.Windows.cs" Condition="'$(TargetsWindows)' == 'true' and '$(FeatureCoreCLR)' != 'true'" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Threading\ThreadPool.Portable.Unix.cs" Condition="'$(TargetsUnix)' == 'true'" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Threading\ThreadPool.Portable.Unix.cs" Condition="'$(TargetsUnix)' == 'true' or '$(TargetsBrowser)' == 'true'" />
```
should this be predicated on FeatureWasmThreads too, like the overlapped one is?

kg commented on src/mono/mono.proj:
```diff
@@ -321,17 +325,25 @@
       <_MonoMinimal Condition="'$(Configuration)' == 'Release'">,debugger_agent,log_dest</_MonoMinimal>
       <_MonoMinimal Condition="'$(Configuration)' == 'Release' and '$(MonoEnableAssertMessages)' != 'true'">$(_MonoMinimal),assert_messages</_MonoMinimal>
+      <_MonoMinimal Condition="'$(MonoWasmThreads)' != 'true'">$(_MonoMinimal),threads</_MonoMinimal>
+      <_MonoMinimal Condition="'$(MonoWasmThreadsNoUser)' == 'true'">$(_MonoMinimal),wasm_user_threads</_MonoMinimal>
```
this looks backwards, i assume ENABLE_MINIMAL=x is a list of features we're *removing*?

lambdageek replied:
> a list of features we're removing

Correct.  It ends up defining `DISABLE_whatever` in the C code.  So in this case, we have `DISABLE_THREADS` if `MonoWasmThreads` isn't set to `true`, and `DISABLE_WASM_USER_THREADS` if `MonoWasmThreadsNoUser` _is_ set to true.

The idea being that we will do a build with `!defined (DISABLE_THREADS) && defined(DISABLE_WASM_USER_THREADS)` for the "EventPipe gets threads and nobody else does" configuration.  That way all of the normal mono threading stuff is just checking for `!defined (DISABLE_THREADS)` (there's already a ton of code under this condition), and the one or two places that are in charge of starting user threads will do an additional check for `!defined (DISABLE_WASM_USER_THREADS)`.

kg commented on src/mono/sample/wasm/wasm.mk:
```diff
@@ -35,7 +35,7 @@ run-browser:
 		exit 1; \
 	else  \
-		$(DOTNET) serve -d:bin/$(CONFIG)/AppBundle $(OPEN_BROWSER) -p:8000; \
+		$(DOTNET) serve -d:bin/$(CONFIG)/AppBundle -h "Cross-Origin-Opener-Policy:same-origin" -h "Cross-Origin-Embedder-Policy:require-corp" $(OPEN_BROWSER) -p:8000; \
```
since this is fairly magical it might be good to have a comment here explaining why we're setting these headers (and doing it in wasm.mk), i.e. (i don't know the syntax for comments in mk files)

```
# we set specific headers to enable SharedArrayBuffer support in browsers for threading: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements
```

kg commented on src/mono/wasm/emscripten-version.txt:
```diff
@@ -1 +1 @@
-3.1.7
\ No newline at end of file
+3.1.7
```
probably worth reverting this since it's a pure whitespace change

kg commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -1323,3 +1336,13 @@ mono_wasm_load_profiler_aot (const char *desc)
+{
+	// At this time we don't use a dedicated thread for finalization even if threading is enabled.
+	// Finalizers periodically run on the main thread
+#if 0
```
if possible let's replace these finalizer related 'if 0's with a common name that we plan to define eventually, so anyone looking for the rest of them can just grep the tree for it instead of having to examine every if 0

lambdageek replied:
Do you think it's even worth keeping?  I have a feeling a lot of wasm-targeting code will want to do finalization on the main thread only, so we might never enable this.

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -293,6 +294,10 @@ function initializeImportsAndExports(
 
     configure_emscripten_startup(module, exportedAPI);
 
+    // HACK: Maintain compatibility with emscripten's generated dotnet.worker.js
```
this is my fault, but we should probably expand this comment before merging it. maybe:

```
// HACK: Emscripten expects the return value of this function to always be the Module object, but
//  we changed ours to return a set of exported namespaces. In order for the emscripten generated
//  worker code to keep working, we detect that we're running in a worker (via the presence of 
//  globalThis.importScripts) and emulate the old behavior.
// Note that this will impact anyone trying to load us in a web worker directly, not just emscripten!
```

kg commented on src/mono/wasm/runtime/method-binding.ts:
```diff
@@ -624,4 +624,4 @@ export type BoundMethodToken = {
     scratchExceptionRoot: WasmRoot<MonoObject>;
-}
\ No newline at end of file
+}
```
maybe revert all the files that have whitespace damage at the end if nothing else changed... i'm not sure where this came from, maybe me originally?

kg commented on src/mono/wasm/wasm.proj:
```diff
@@ -125,10 +134,14 @@
 
     <ItemGroup>
       <_EmccLinkFlags Include="-s EXPORT_ES6=1" Condition="'$(WasmEnableES6)' == 'true'" />
+      <_EmccLinkFlags Include="-s LLD_REPORT_UNDEFINED" />
```
intended?

lambdageek replied:
Looking at the Emscripten release notes, we probably don't want this on by default.  @steveisok what was the intention here?  Just validating that all the EP stuff got built/linked?

steveisok replied:
It was used to potentially root out issues.  We can remove it or if it makes sense, enable by flag.

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -26,6 +26,29 @@ export const mono_wasm_runtime_is_initialized = new Promise((resolve, reject) =>
+        console.log("determined url of main script to be " + temp.href);
+        (<any>module)["mainScriptUrlOrBlob"] = temp.href;
+      }
+    } else
+        console.log("could not determine url of main script for workers");
```
This is too chatty.  And also it complains in v8 runs.

lambdageek commented on src/mono/wasm/wasm.proj:
```diff
@@ -125,10 +134,14 @@
       <_EmccLinkFlags Include="-s EXPORTED_RUNTIME_METHODS=&quot;['FS','print','ccall','cwrap','setValue','getValue','UTF8ToString','UTF8ArrayToString','FS_createPath','FS_createDataFile','removeRunDependency','addRunDependency', 'FS_readFile']&quot;"  />
+      <_EmccCommonFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s USE_PTHREADS=1" />
+      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-Wno-pthreads-mem-growth" />
+      <_EmccLinkFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s PTHREAD_POOL_SIZE=2" />
```
We should probably make this configurable somehow.  Otherwise code like this will deadlock because the main thread never returns to the JS eventloop, so Emscripten can't start the third thread once it exhausts its initial pool:

```csharp
public static void Main() {
   Thread[] threads = new Thread[] { new Thread(ThreadMain1), new Thread(ThreadMain2) }
   threads[0].Start();
   threads[1].Start();
   threads[0].Join();
   threads[1].Join();
}
```

steveisok replied:
The [docs](https://emscripten.org/docs/porting/pthreads.html#compiling-with-pthreads-enabled) indicate that `PTHREAD_POOL_SIZE` can be an integer or a javascript expression.  `navigator.hardwareConcurrency` might be a  better default.

Emscripten has some built in detection when the thread pool is exhausted.  `PTHREAD_POOL_SIZE_STRICT` by default warns when the thread pool is exhausted.  It can also be adjusted to error (`PTHREAD_POOL_SIZE_STRICT=2`) or do nothing (`PTHREAD_POOL_SIZE_STRICT=0`).

lambdageek replied:
Made an issue to revisit this https://github.com/dotnet/runtime/issues/68397

lambdageek committed:
remove unneeded configure checks

lambdageek committed:
revert line damage; add better comment in export.ts

lambdageek committed:
Exclude Microsoft.NET.WebAssembly.Threading from testPackages.proj

lambdageek committed:
review feedback

lambdageek commented:
The wasm AOT failures seem relevant. I will be taking a look at those.

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/Directory.Build.props:
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
+    <!-- Disable all binplacing since this assembly will clash with the real ref assemblies -->
+    <EnableBinPlacing>false</EnableBinPlacing>
+    <StrongNameKeyId>Microsoft</StrongNameKeyId>
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
```
You shouldn't need any of these properties as projects in this directory don't compile. I believe you can believe this entire file.

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <PackageDescription>Exposes Threading APIs for WebAssembly projects</PackageDescription>
+    <PackageId>$(MSBuildProjectName)</PackageId>
```
PackageId already defaults to the AssemblyName which defaults to the project name: https://github.com/dotnet/msbuild/blob/3ade6423189769545ddff2ffeeed37010ec57f4d/src/Tasks/Microsoft.Common.CurrentVersion.targets#L238

```suggestion
```

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+    <DebugSymbols>false</DebugSymbols>
+    <DebugType>none</DebugType>
+    <NoTargetsDoNotReferenceOutputAssemblies>false</NoTargetsDoNotReferenceOutputAssemblies>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
```
This property gets set automatically in packaging.targets: https://github.com/dotnet/runtime/blob/main/eng/packaging.targets#L18-L26.

```suggestion
```

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <!-- XXX FIXME: the rest of these props affect the NoTargets SDK in some way.  I want to be able to reference this project from another .csproj -->
+    <NoTargetsDoNotReferenceOutputAssemblies>false</NoTargetsDoNotReferenceOutputAssemblies>
+    <DisableImplicitFrameworkReferences>false</DisableImplicitFrameworkReferences>
```
As this project doesn't compile anything this setting should be removed. Otherwise you would require the either live or prebuilt targeting pack to be available which should be avoided (as the project doesn't compile).

```suggestion
```

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+    <NoWarn>$(NoWarn);NU5128;NU5131</NoWarn>
+    <!-- Place assemblies in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <!-- XXX FIXME: the rest of these props affect the NoTargets SDK in some way.  I want to be able to reference this project from another .csproj -->
```
What's the reason that you want to reference this project from another csproj? NoTargets SDK projects aren't really useful to be referenced by other projects (except to do the packaging work or invoke custom targets). NoTargets projects don't even return a target path as they don't compile.

ViktorHofer replied:
Yeah, for that you would need a Microsoft.Build.Traversal project but that isn't packable so you are better off with your current strategy to directly reference the ref projects. I assume you can delete this block then.

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Threading.Thread.WebAssembly.Threading\ref\System.Threading.Thread.WebAssembly.Threading.csproj" Pack="true" PrivateAssets="all" Private="true" IncludeReferenceAssemblyInPackage="true" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Threading.ThreadPool.WebAssembly.Threading\ref\System.Threading.ThreadPool.WebAssembly.Threading.csproj" Pack="true" PrivateAssets="all" Private="true" IncludeReferenceAssemblyInPackage="true" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Diagnostics.Tracing.WebAssembly.PerfTracing\ref\System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj" Pack="true" PrivateAssets="all" Private="true" IncludeReferenceAssemblyInPackage="true" />
```
As all these ProjectReferences have the same attributes, consider using only one item declaration instead:

```suggestion
    <ProjectReference Include="$(LibrariesProjectRoot)System.Threading.Thread.WebAssembly.Threading\ref\System.Threading.Thread.WebAssembly.Threading.csproj;$(LibrariesProjectRoot)System.Diagnostics.Tracing.WebAssembly.PerfTracing\ref\System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj;$(LibrariesProjectRoot)System.Threading.ThreadPool.WebAssembly.Threading\ref\System.Threading.ThreadPool.WebAssembly.Threading.csproj" Pack="true" PrivateAssets="all" Private="true" IncludeReferenceAssemblyInPackage="true" />
```

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,47 @@
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Diagnostics.Tracing.WebAssembly.PerfTracing\ref\System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj" Pack="true" PrivateAssets="all" Private="true" IncludeReferenceAssemblyInPackage="true" />
+  </ItemGroup>
+
+  <!-- https://github.com/nuget/home/issues/3891 add the references .dlls to the package, and don't add the to the nuspec -->
+  <Target Name="CopyProjectReferencesToPackage" DependsOnTargets="BuildOnlySettings;ResolveReferences">
+    <ItemGroup>
+      <BuildOutputInPackage Include="@(ReferenceCopyLocalPaths->WithMetadataValue('ReferenceSourceTarget', 'ProjectReference')->WithMetadataValue('PrivateAssets', 'All'))" />
+    </ItemGroup>
+  </Target>
```
Hmm, why do you need this extra target? https://github.com/lambdageek/runtime/blob/eed9c390017a5a52c2b0ce7a6be5deae0e8660ac/eng/packaging.targets#L233-L258 should already do all that for you.

lambdageek replied:
Yea I think I can delete this. It's leftover from trying to get both `PackageReference`  and `ProjectReference` working.   Since I gave up on `ProjectReference`, I don't think this is needed anymore.

ViktorHofer commented on src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt:
```diff
@@ -145,3 +145,4 @@ CannotRemoveAttribute : Attribute 'System.Runtime.Versioning.UnsupportedOSPlatfo
 Compat issues with assembly System.Security.Cryptography.X509Certificates:
 CannotChangeAttribute : Attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute' on 'System.Security.Cryptography.X509Certificates.PublicKey.GetDSAPublicKey()' changed from '[UnsupportedOSPlatformAttribute("ios")]' in the contract to '[UnsupportedOSPlatformAttribute("browser")]' in the implementation.
 Total Issues: 136
+
```
Please revert:

```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/src/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.Build.Traversal">
```
What's the reason to have this file? Packaging doesn't require it.

lambdageek replied:
Without this, I wasn't getting a nuget in `artifacts/packages/` if I just did `./build.sh --os browser -c Release` (both with or without `/p:WasmEnableThreading`).

ViktorHofer replied:
I don't think you need these Traversal projects at all anymore. In the past they were required but they shouldn't be anymore.

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/ref/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,37 @@
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
+    <IsPackable>false</IsPackable>
```
That's already the default setting.

```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/src/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.Build.Traversal">
```
Same comment. I don't think you need that file.

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,39 @@
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
+    <IsPackable>false</IsPackable>
```
Same as above, that's already the default.

```suggestion
```

ViktorHofer commented on src/libraries/System.Diagnostics.Tracing.WebAssembly.PerfTracing/ref/System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj:
```diff
@@ -0,0 +1,33 @@
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <DefineConstants>$(DefineConstants);FEATURE_WASM_PERFTRACING</DefineConstants>
+    <IsPackable>false</IsPackable>
```
```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,39 @@
+    <ProjectReference Include="..\..\System.Security.Principal\ref\System.Security.Principal.csproj" PrivateAssets="all" />
+    <ProjectReference Include="..\..\System.Threading\ref\System.Threading.csproj" PrivateAssets="all" />
+  </ItemGroup>
+  <Target Name="_ExperimentalUpdateFileVersion"
+          AfterTargets="_InitializeAssemblyVersion">
+    <PropertyGroup>
+      <_FileVersionMaj>$(FileVersion.Split('.')[0])</_FileVersionMaj>
+      <_FileVersionMin>$(FileVersion.Split('.')[1])</_FileVersionMin>
+      <_FileVersionBld>$(FileVersion.Split('.')[2])</_FileVersionBld>
+      <_FileVersionRev>$(FileVersion.Split('.')[3])</_FileVersionRev>
+      <FileVersion>$(_FileVersionMaj).$([MSBuild]::Add($(_FileVersionMin), 100)).$(_FileVersionBld).$(_FileVersionRev)</FileVersion>
+    </PropertyGroup>
+  </Target>
```
You have this target in multiple projects. Please consider moving it into packaging.targets instead and i.e. condition it on a property that you can set in each project.

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/Directory.Build.props:
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
+    <!-- Disable all binplacing since this assembly will clash with the real System.Runtime -->
+    <EnableBinPlacing>false</EnableBinPlacing>
+    <StrongNameKeyId>Microsoft</StrongNameKeyId>
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
```
Directory.Build.props files only make sense if the directory contains multiple projects (that compile). I would delete all these files and move these three properties directly into the ref projects.

ViktorHofer commented on src/libraries/testPackages/testPackages.proj:
```diff
@@ -29,6 +29,8 @@
 
     <!-- runtime.native.System.IO.Ports requires all runtime legs to be built in order to test -->
     <ExcludePackages Include="runtime.native.System.IO.Ports" />
+    <!-- Microsoft.NET.WebAssembly.Threading only has ref assemblies, no runtime items -->
+    <ExcludePackages Include="Microsoft.NET.WebAssembly.Threading" />
```
@ericstj Should ref only packages be excluded from package testing?

lambdageek replied:
The package testing actually complained "you don't have any runtime items in this package" so it would be nice to have a way to say "ensure that I don't have any runtime items in this package":
```
D:\a\_work\1\s\artifacts\bin\testPackages\packageTest.targets(94,5): error : Could not locate any runtime items from Package Microsoft.NET.WebAssembly.Threading [D:\a\_work\1\s\artifacts\bin\testPackages\projects\Microsoft.NET.WebAssembly.Threading\project.csproj]
```

ViktorHofer replied:
Presumably by setting `SkipVerifyRuntime` to true in a file under https://github.com/dotnet/runtime/tree/main/src/libraries/testPackages/packageSettings.

ericstj replied:
> @ericstj Should ref only packages be excluded from package testing

Yeah -- I wouldn't have expected we'd have many `ref` only packages though we've seemed to found new ways to use them to do interesting experiments.  If that continues maybe we should simplify the pattern.

lambdageek commented on src/libraries/System.Threading.ThreadPool/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1,11 @@
+Compat issues with assembly System.Threading.ThreadPool:
+CannotRemoveAttribute : Attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute' exists on 'System.Threading.RegisteredWaitHandle' in the contract but not the implementation.
```
No, it's the other way around. 

* CoreLib does not have the attribute (it checks the `Thread.IsThreadStartSupported` flag at runtime).  

* The non-experimental ref assembly always has the attribute.  (So there is always a CannotRemoveAttribute issued)

* The experimental ref  assembly does not have the attribute. (So here the implementation matches the contract)

ViktorHofer replied:
> The non-experimental ref assembly always has the attribute. (So there is always a CannotRemoveAttribute issued)

Why wasn't that baseline required before then? Because the implementation had it before at build time and now it doesn't anymore?

lambdageek replied:
> > The non-experimental ref assembly always has the attribute. (So there is always a CannotRemoveAttribute issued)
> 
> Why wasn't that baseline required before then? Because the implementation had it before at build time and now it doesn't anymore?

yes. previously both the implementation and the ref assembly had the attribute

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

lambdageek committed:
Use a testPackages settings file to skip package runtime item verification

lambdageek committed:
remove SIZEOF_VOID_P == 7 check

lambdageek committed:
remove unneeded Directory.Build.props for ref package

since ti doesn't compile its own assembly, none of these properties are needed

lambdageek committed:
use one ProjectReference item to share metadata for the ref assemblies

lambdageek committed:
remove ProjectReference comment and NoTargetsDoNotReferenceOutputAssemblies prop

lambdageek committed:
Remove unneeded target

lambdageek committed:
Merge branch 'main' into wasm-threading-eventpipe

lambdageek committed:
packaging simplification

- move `_ExperimentalUpdateFileVersion` target to packaging.targets, conditioned on a
new `IsExperimentalRefAssembly` attribute.  (The target increments the file
version of the ref assembly to make it easier to distinguish from the real
non-experimental ref assembly)
- Remove unneeded src subdirectories in ref assembly libraries
- Move properties that are only used in the ref assembly projects directory
into the projects and delete Directory.Build.props in the experimental ref
assembly subdirectories.

ViktorHofer commented on eng/packaging.targets:
```diff
@@ -269,4 +269,20 @@
+       assembly), bump their minor file version by 100 to make it distinguishable from the regular
+       ref assembly.
+  -->
+  <Target Name="_ExperimentalRefAssemblyUpdateFileVersion"
```
nit: We don't use lowercase prefixes when a target isn't just an implementation detail:

```suggestion
  <Target Name="UpdateExperimentalRefAssemblyFileVersion"
```

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,34 @@
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(FeatureWasmThreads)' == 'true'">
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
```
This library isn't packable.

```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,34 @@
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
+    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
```
This library isn't packable.

```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/ref/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,32 @@
+    <!-- Place this assembly in the nuget ref folder -->
+    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
+    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
+    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
```
This library isn't packable.

```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/ref/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,32 @@
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(FeatureWasmThreads)' == 'true'">
+    <!-- Place this assembly in the nuget ref folder -->
```
This library isn't packable.

```suggestion
```

ViktorHofer commented on src/libraries/Microsoft.NET.WebAssembly.Threading/src/Microsoft.NET.WebAssembly.Threading.proj:
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <PackageDescription>Exposes Threading APIs for WebAssembly projects</PackageDescription>
```
small nit: Can you please move that line down after IsPacakble to group packaging properties together?

ViktorHofer commented:
@lambdageek you are on the right track :) Just some remaining feedback.

lambdageek committed:
move and rename UpdateExperimentalRefAssemblyFileVersion target

packages.targets is only included for IsPackable=true projects, and these ref
assemblies are not packable.

lambdageek committed:
Assorted code review nits

ViktorHofer commented on src/libraries/System.Diagnostics.Tracing.WebAssembly.PerfTracing/ref/System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj:
```diff
@@ -1,22 +1,19 @@
-    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
-    <DefineConstants>$(DefineConstants);FEATURE_WASM_PERFTRACING</DefineConstants>
-    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
+  <PropertyGroup>
```
nit: merge these two property groups.

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -1,26 +1,19 @@
-    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
-    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
-    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
+  <PropertyGroup>
```
nit: merge these two property groups.

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/ref/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -2,25 +2,18 @@
-    <BuildOutputTargetFolder>ref</BuildOutputTargetFolder>
-    <DefineConstants>$(DefineConstants);FEATURE_WASM_THREADS</DefineConstants>
-    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
+  <PropertyGroup>
```
nit: merge these two property groups

ViktorHofer commented on src/libraries/System.Threading.ThreadPool/ref/System.Threading.ThreadPool.csproj:
```diff
@@ -3,6 +3,10 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
     <Nullable>enable</Nullable>
+    <FeatureWasmThreads>false</FeatureWasmThreads>
```
nit: do you need this to be set to false? The above condition only checks on true and that's usually sufficient in the msbuild world.

lambdageek replied:
I think I like it as documentation and a reminder to flip it over to `true` once this stuff is no longer experimental, or once we have a [better comptability analyzer](https://github.com/dotnet/designs/pull/261)

ViktorHofer commented on src/libraries/System.Threading.ThreadPool/ref/System.Threading.ThreadPool.csproj:
```diff
@@ -3,6 +3,10 @@
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
     <Nullable>enable</Nullable>
+    <FeatureWasmThreads>false</FeatureWasmThreads>
+  </PropertyGroup>
+  <PropertyGroup>
```
```suggestion
```

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -0,0 +1,27 @@
+    <EnableBinPlacing>false</EnableBinPlacing>
+    <StrongNameKeyId>Microsoft</StrongNameKeyId>
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
+  </PropertyGroup>
+  <PropertyGroup>
```
```suggestion
  </PropertyGroup>
  <PropertyGroup>
```

ViktorHofer commented on src/libraries/System.Diagnostics.Tracing/ref/System.Diagnostics.Tracing.csproj:
```diff
@@ -3,6 +3,10 @@
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
     <Nullable>enable</Nullable>
+    <FeatureWasmPerfTracing>false</FeatureWasmPerfTracing>
+  </PropertyGroup>
+  <PropertyGroup>
```
```suggestion
```

ViktorHofer commented on src/libraries/System.Diagnostics.Tracing.WebAssembly.PerfTracing/ref/System.Diagnostics.Tracing.WebAssembly.PerfTracing.csproj:
```diff
@@ -0,0 +1,25 @@
+    <EnableBinPlacing>false</EnableBinPlacing>
+    <StrongNameKeyId>Microsoft</StrongNameKeyId>
+    <IncludePlatformAttributes>true</IncludePlatformAttributes>
+  </PropertyGroup>
+  <PropertyGroup>
```
```suggestion
```

lambdageek committed:
Don't build/pack the multi-threaded sample on single-threaded runtime

lambdageek committed:
remove gratuitous debug printfs

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-threading-eventpipe

lambdageek committed:
merge followup: nullable is enabled by default now

lambdageek committed:
make eslint happy

lambdageek committed:
include wasm-config.h in wasm runtime host

lambdageek committed:
add wasm-config.h directory to includde directories

lambdageek committed:
include wasm-config.h into the runtime pack

fixes aot compilation

steveisok reviewed (approved):
Great job putting this together!

lambdageek committed:
Add wasm-config.h to manifest

ViktorHofer reviewed (approved):
Infrastructure changes LGTM. Good work 👍

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
put wasm-config.h into include/wasm from the outset

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
put back noExitRuntime replacement for CJS

lateralusX commented on src/mono/mono/eventpipe/CMakeLists.txt:
```diff
@@ -10,9 +10,9 @@ if(ENABLE_PERFTRACING)
         add_definitions(-DENABLE_PERFTRACING_PAL_TCP)
     endif (FEATURE_PERFTRACING_PAL_TCP)
 
-    if (FEATURE_PERFTRACING_DISABLE_LISTEN_PORTS)
```
Why was this needed?

lambdageek replied:
It's a typo, the old cmake variable name was unused

lateralusX commented on src/mono/wasm/threads.md:
```diff
@@ -0,0 +1,80 @@
+
+## Building ##
+
+Build with `/p:WasmEnableThreads=true` to enable support for multi-threading.
```
Will this also enable EventPipe diagnostics support?

lambdageek replied:
I think so.  Once we have full user-visible multithreading there's no particular reason to exclude eventpipe.

lateralusX reviewed (approved):
LGTM!

akoeplinger reviewed (commented):
LGTM thank you! Found just one small formatting nit.

lewing commented:
🎊
